internal/audio/adaptive_buffer.go

```diff
@@ -0,0 +1,338 @@
+package audio
+
+import (
+	"context"
+	"math"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"github.com/jetkvm/kvm/internal/logging"
+	"github.com/rs/zerolog"
+)
+
+// AdaptiveBufferConfig holds configuration for adaptive buffer sizing
+type AdaptiveBufferConfig struct {
+	// Buffer size limits (in frames)
+	MinBufferSize     int
+	MaxBufferSize     int
+	DefaultBufferSize int
+
+	// System load thresholds
+	LowCPUThreshold     float64 // Below this, increase buffer size
+	HighCPUThreshold    float64 // Above this, decrease buffer size
+	LowMemoryThreshold  float64 // Below this, increase buffer size
+	HighMemoryThreshold float64 // Above this, decrease buffer size
+
+	// Latency thresholds (in milliseconds)
+	TargetLatency time.Duration
+	MaxLatency    time.Duration
+
+	// Adaptation parameters
+	AdaptationInterval time.Duration
+	SmoothingFactor    float64 // 0.0-1.0, higher = more responsive
+}
+
+// DefaultAdaptiveBufferConfig returns optimized config for JetKVM hardware
+func DefaultAdaptiveBufferConfig() AdaptiveBufferConfig {
+	return AdaptiveBufferConfig{
+		// Conservative buffer sizes for 256MB RAM constraint
+		MinBufferSize:     3,  // Minimum 3 frames (slightly higher for stability)
+		MaxBufferSize:     20, // Maximum 20 frames (increased for high load scenarios)
+		DefaultBufferSize: 6,  // Default 6 frames (increased for better stability)
+
+		// CPU thresholds optimized for single-core ARM Cortex A7 under load
+		LowCPUThreshold:  20.0, // Below 20% CPU
+		HighCPUThreshold: 60.0, // Above 60% CPU (lowered to be more responsive)
+
+		// Memory thresholds for 256MB total RAM
+		LowMemoryThreshold:  35.0, // Below 35% memory usage
+		HighMemoryThreshold: 75.0, // Above 75% memory usage (lowered for earlier response)
+
+		// Latency targets
+		TargetLatency: 20 * time.Millisecond, // Target 20ms latency
+		MaxLatency:    50 * time.Millisecond, // Max acceptable 50ms
+
+		// Adaptation settings
+		AdaptationInterval: 500 * time.Millisecond, // Check every 500ms
+		SmoothingFactor:    0.3,                    // Moderate responsiveness
+	}
+}
+
+// AdaptiveBufferManager manages dynamic buffer sizing based on system conditions
+type AdaptiveBufferManager struct {
+	// Atomic fields MUST be first for ARM32 alignment (int64 fields need 8-byte alignment)
+	currentInputBufferSize  int64 // Current input buffer size (atomic)
+	currentOutputBufferSize int64 // Current output buffer size (atomic)
+	averageLatency          int64 // Average latency in nanoseconds (atomic)
+	systemCPUPercent        int64 // System CPU percentage * 100 (atomic)
+	systemMemoryPercent     int64 // System memory percentage * 100 (atomic)
+	adaptationCount         int64 // Metrics tracking (atomic)
```

Comment:
All of these should be `atomic.Uint32` so the atomic behavior is guaranteed (available since Go 1.19)